core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java

```diff
@@ -0,0 +1,578 @@
+        throw new UnsupportedOperationException("Unsupported delete file type: " + file.content());
+    }
+  }
+
+  private static <F extends ContentFile<F>> void appendEntryWithFile(
+      ManifestEntry<F> entry, ManifestWriter<F> writer, F file) {
+
+    switch (entry.status()) {
+      case ADDED:
+        writer.add(file);
+        break;
+      case EXISTING:
+        writer.existing(
+            file, entry.snapshotId(), entry.dataSequenceNumber(), entry.fileSequenceNumber());
+        break;
+      case DELETED:
+        writer.delete(file, entry.dataSequenceNumber(), entry.fileSequenceNumber());
+        break;
+    }
+  }
+
+  private static DeleteFile newEqualityDeleteEntry(
+      DeleteFile file, PartitionSpec spec, String sourcePrefix, String targetPrefix) {
+    String path = file.location();
+
+    if (!path.startsWith(sourcePrefix)) {
+      throw new UnsupportedOperationException(
+          "Expected delete file to be under the source prefix: "
+              + sourcePrefix
+              + " but was "
+              + path);
+    }
+    int[] equalityFieldIds = file.equalityFieldIds().stream().mapToInt(Integer::intValue).toArray();
+    String newPath = newPath(path, sourcePrefix, targetPrefix);
+    return FileMetadata.deleteFileBuilder(spec)
+        .ofEqualityDeletes(equalityFieldIds)
+        .copy(file)
+        .withPath(newPath)
+        .withSplitOffsets(file.splitOffsets())
+        .build();
+  }
+
+  /** Class providing engine-specific methods to read and write position delete files. */
+  public interface PositionDeleteReaderWriter extends Serializable {
+    CloseableIterable<Record> reader(InputFile inputFile, FileFormat format, PartitionSpec spec);
+
+    PositionDeleteWriter<Record> writer(
+        OutputFile outputFile,
+        FileFormat format,
+        PartitionSpec spec,
+        StructLike partition,
+        Schema rowSchema)
+        throws IOException;
+  }
+
+  /**
+   * Rewrite a position delete file, replacing path references.
+   *
+   * @param deleteFile source delete file to be rewritten
+   * @param outputFile output file to rewrite delete file to
+   * @param io file io
+   * @param spec spec of delete file
+   * @param sourcePrefix source prefix that will be replaced
+   * @param targetPrefix target prefix to replace it
+   * @param posDeleteReaderWriter class to read and write position delete files
+   */
+  public static void rewritePositionDeleteFile(
+      DeleteFile deleteFile,
+      OutputFile outputFile,
+      FileIO io,
+      PartitionSpec spec,
+      String sourcePrefix,
+      String targetPrefix,
+      PositionDeleteReaderWriter posDeleteReaderWriter)
+      throws IOException {
+    String path = deleteFile.location();
+    if (!path.startsWith(sourcePrefix)) {
+      throw new UnsupportedOperationException(
+          "Expected delete file to be under the source prefix: "
+              + sourcePrefix
+              + " but was "
+              + path);
+    }
```

Comment:
nit, how about `String.format("Expected delete file %s to start with prefix: %s", path, sourcePrefix)`